### ISSUE TITLE
--experimental_output_directory_naming_scheme=diff_against_baseline caused Bazel to rebuild whenever --run_under or --test_env change

### ISSUE THREAD

**Issue by linzhp** [bazelbuild/bazel#18480]
### Description of the bug:

When `--experimental_output_directory_naming_scheme=diff_against_baseline`, running `--run_under echo` or `--test_env FOO=bar` would invalidate the cache and rebuild everything. Bazel wouldn't rebuild if `--experimental_output_directory_naming_scheme=diff_against_baseline` is not given.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```bash
git clone git@github.com:bazelbuild/bazel-gazelle.git
cd bazel-gazelle
# populate the cache
bazel build --experimental_output_directory_naming_scheme=diff_against_baseline //cmd/gazelle
# This would rebuild everything.
bazel run --run_under echo --experimental_output_directory_naming_scheme=diff_against_baseline //cmd/gazelle
# This also rebuilds everything
bazel build --test_env FOO=bar --experimental_output_directory_naming_scheme=diff_against_baseline //cmd/gazelle
```

### Which operating system are you running Bazel on?

macOS and Linux

### What is the output of `bazel info release`?

release 6.2.0-homebrew

**Comment by fmeum** [comment]
Cc @gregestren @sdtwigg

**Comment by fmeum** [comment]
`--run_under` is reset in the exec configuration (https://cs.opensource.google/bazel/bazel/+/refs/heads/master:src/main/java/com/google/devtools/build/lib/analysis/config/CoreOptions.java;l=973;drc=93cd68dd788b62c9a51244fd475b456913e44476;bpv=1;bpt=1), which makes me think this would be solved by https://github.com/bazelbuild/bazel/pull/18002.

**Comment by gregestren** [comment]
@sdtwigg for priority triage.

**Comment by sdtwigg** [comment]
Do you still see this behavior when using `diff_against_dynamic_baseline`?

This is a bit peculiar as I would not expected a difference in analysis behavior due to that flag. In particular changes to either `run_under` and `test_env` would invalidate the analysis cache regardless of the setting of `experimental_output_directory_naming_scheme`.

Are you consistently seeing analysis cache discards but a difference in action cache behavior? If so, changing to `diff_against_dynamic_baseline` should fix the issue. (Technically, only for actions after an exec transition but actions before one should not have been exhibiting that issue anyways.)

**Comment by linzhp** [comment]
#18561 fixes this issue

**Comment by fmeum** [comment]
@sdtwigg Based on https://github.com/bazelbuild/bazel/issues/18480#issuecomment-1684669608, could we proceed with the merge of #18561?

**Comment by sdtwigg** [comment]
Yes, this looks all together now and was hoping to do the merge today or so.

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by fmeum** [comment]
Cherry-picked in https://github.com/bazelbuild/bazel/pull/19514

**Comment by iancha1992** [comment]
@bazel-io fork 6.4.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 6.4.0 RC1](https://releases.bazel.build/6.4.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
